pageserver/src/pgdatadir_mapping.rs

```diff
@@ -252,14 +252,8 @@ impl Timeline {
         let mut buf = version.get(self, key, ctx).await?;
         let nblocks = buf.get_u32_le();
 
-        if latest {
-            // Update relation size cache only if "latest" flag is set.
-            // This flag is set by compute when it is working with most recent version of relation.
-            // Typically master compute node always set latest=true.
-            // Please notice, that even if compute node "by mistake" specifies old LSN but set
-            // latest=true, then it can not cause cache corruption, because with latest=true
-            // pageserver choose max(request_lsn, last_written_lsn) and so cached value will be
-            // associated with most recent value of LSN.
+        if horizon == Lsn::MAX {
+            // Update relation size cache only if latest version is requested.
             self.update_cached_rel_size(tag, version.get_lsn(), nblocks);
```

Comment:
The old comment for why this is safe is still valid, right?

Reply:
How significant is it to update the cache on reads anyway? I'd be inclined to just get rid of this, unless it makes a noticeable difference. I'm not 100% convinced the that reasoning is correct, and at least it depends on another assumption that hasn't been mentioned: it assumes that we haven't evicted any entries from the cache, since the WAL ingestion started. That currently holds, but at least we should document that somewhere.

Also, we should add a comment about that assumption about choosing max(request_lsn, last_written_lsn) also in wait_or_get_last_lsn, where we do that.  (BTW, shouldn't it be last_record_lsn rather than last_written_lsn?). We might actually want to change that, so that wait_or_get_last_lsn would use an older LSN if it can, because that might be more performant. Using an older LSN when digging through the layers could allow skipping the in-memory layer and recent L0 layers altogether, for example.

Reply:
Here's a PR to refactor that part, so that we can eliminate those assumptions: https://github.com/neondatabase/neon/pull/7376. It's pretty orthogonal to this work, but would be nice to tighten that up.